tests/hydra.rs

```diff
@@ -440,3 +452,209 @@ fn scenario_2() -> TestResult {
+fn hydra_oura_stdout_test(file: String, expected: Vec<String>) -> TestResult {
+    let rt = Runtime::new().unwrap();
+    let (tx, _rx) = mpsc::channel();
+    let _ = rt.block_on(async move {
+        let addr = "127.0.0.1:4001".to_string();
+        let server = TcpListener::bind(&addr).await?;
+        println!("WebSocket server started on ws://{}", addr);
+
+        let _ = tokio::spawn(async move { oura_pipeline(expected).await });
+
+        while let Ok((stream, _)) = server.accept().await {
+            tokio::spawn(handle_connection(stream, file, tx));
+            time::sleep(Duration::from_secs(3)).await;
+            break;
+        }
+
+        Ok::<(), std::io::Error>(())
+    });
+    Ok(())
+}
+
+async fn handle_connection(
+    stream: tokio::net::TcpStream,
+    file: String,
+    tx: mpsc::Sender<usize>,
+) -> Result<()> {
+    let mut ws_stream = accept_async(stream).await?;
+    println!("WebSocket server oura connection established");
+
+    let to_send = fs::read_to_string(file)?;
+
+    let mut lines = 0;
+    for line in to_send.lines() {
+        ws_stream.send(Message::Text(line.to_string())).await?;
+        lines += 1;
+    }
+    tx.send(lines).unwrap();
+
+    Ok(())
+}
+
+async fn websocket_client(expected: String, rx: mpsc::Receiver<usize>) -> Result<()> {
+    let url = Url::parse("ws://127.0.0.1:4001")?;
+    let (mut ws_stream, _) = connect_async(url.as_str())
+        .await
+        .expect("Failed to connect");
+    println!("WebSocket client oura connected");
+
+    let mut received = vec![];
+    let mut msgs_number = 0;
+
+    // Receiving messages from the server
+    while let Some(msg) = ws_stream.next().await {
+        match msg? {
+            Message::Text(text) => {
+                received.push(text);
+                //println!("WebSocket client received message from server: {}", text);
+            }
+            _ => {}
+        }
+        if let Ok(size) = rx.try_recv() {
+            msgs_number = size;
+        }
+        if msgs_number == received.len() {
+            break;
+        }
+    }
+
+    let mut joined = received.join("\n");
+    joined.push_str("\n");
+
+    assert_eq!(joined, expected);
+    println!("WebSocket client oura disconnected");
+
+    Ok(())
+}
+
+async fn oura_pipeline(expected: Vec<String>) -> Result<()> {
+    println!("oura_pipeline online");
+
+    //Clean output file
+    let _ = std::process::Command::new("truncate")
+        .arg("-s 0")
+        .arg("tests/hydra/logs.txt")
+        .spawn();
+
+    tokio::spawn(invoke_pipeline());
+    time::sleep(Duration::from_secs(1)).await;
+
+    let jsons = fs::read_to_string("tests/hydra/logs.txt")?;
+    let mut predicates = vec![];
+    let mut count = 0;
+    for json in jsons.lines() {
+        let predicate_fn = predicate::str::contains(&expected[count]);
+        predicates.push(predicate_fn.eval(json));
+        count += 1;
+    }
+    assert_eq!(predicates, vec![true; expected.len()]);
```

Comment:
Off the top of my head, I find it odd that the `oura_pipeline` is responsible for both capturing the output and performing the assertion. Did you consider returning the captured outputs in the result? 

And — maybe we could check the output with something like [goldenfile](https://crates.io/crates/goldenfile) instead of `assert_eq` if the output is deterministic? 🤔

Reply:
added golden as suggested :+1: